### ISSUE TITLE
Intermittent "TypeError: Cannot read property 'pop' of undefined"

### ISSUE THREAD

**Issue by PeterBorah** [ConsenSys-archive/ganache-cli-archive#81]
I run my truffle tests against testrpc many times a day, and maybe once every 20 runs or so, this happens:

```
/usr/local/lib/node_modules/ethereumjs-testrpc/node_modules/merkle-patricia-tree/baseTrie.js:359
  var lastNode = stack.pop()
                      ^

TypeError: Cannot read property 'pop' of undefined
    at CheckpointTrie.Trie._updateNode (/usr/local/lib/node_modules/ethereumjs-testrpc/node_modules/merkle-patricia-tree/baseTrie.js:359:23)
    at /usr/local/lib/node_modules/ethereumjs-testrpc/node_modules/merkle-patricia-tree/baseTrie.js:106:16
    at /usr/local/lib/node_modules/ethereumjs-testrpc/node_modules/merkle-patricia-tree/baseTrie.js:442:14
    at processNode (/usr/local/lib/node_modules/ethereumjs-testrpc/node_modules/merkle-patricia-tree/baseTrie.js:447:23)
    at /usr/local/lib/node_modules/ethereumjs-testrpc/node_modules/merkle-patricia-tree/baseTrie.js:485:11
    at /usr/local/lib/node_modules/ethereumjs-testrpc/node_modules/merkle-patricia-tree/baseTrie.js:179:7
    at /usr/local/lib/node_modules/ethereumjs-testrpc/node_modules/merkle-patricia-tree/util.js:75:7
    at /usr/local/lib/node_modules/ethereumjs-testrpc/node_modules/async/lib/async.js:52:16
    at /usr/local/lib/node_modules/ethereumjs-testrpc/node_modules/async/lib/async.js:269:32
    at /usr/local/lib/node_modules/ethereumjs-testrpc/node_modules/async/lib/async.js:44:16

```

I haven't noticed any patterns about when this happens; it seems random to me.

**Comment by Schaeff** [comment]
Same here!

**Comment by tcoulter** [comment]
I haven't been able to reproduce this, but it's an intermittent failures
somewhere in the bowels of EthereumJS (looks like the merkle tree). My
hunch is that when a transaction is being saved some array that's supposed
to exist doesn't -- unfortunately I have nothing more to go on. If one of
you can find the smallest example that exposes the issue, that'd be very
helpful in debugging what's going on!

On Thu, May 12, 2016 at 6:38 AM, Thibaut SCHAEFFER <notifications@github.com

**Comment by chevdor** [comment]
I ran into the issue a few times today using truffle beta v3.0.0 but I did not spot yet a pattern to help reproduce the issue.

**Comment by area** [comment]
I haven't been seeing this error since using a custom version of `testrpc` that disables the snapshot and revert endpoints. If other people can confirm that behaviour, that'd at least point us a little towards what's causing this.

**Comment by chevdor** [comment]
I am using testrpc 3.0.0 and truffle thru the webpack solidity loader now and I must say that I did not see this error in a while.

**Comment by PeterBorah** [comment]
I still see it fairly regularly. 

@area: How might I go about verifying that behavior?

**Comment by area** [comment]
If you're using testrpc 3.0.0, I have a branch that only differs in that those endpoints are switched off, so truffle doesn't use them.

```
npm uninstall -g ethereumjs-testrpc
npm install -g https://github.com/area/testrpc.git#disable/revert
```

**Comment by mhhf** [comment]
I've recently implemented persistency into testrpc and run excessively into this bug and maybe can shine some light on where this comes from: The troublemaker is levelup + memdown which is used as an in memory data storage for the merkle-patricia-tree **and** other data.
levelup has the property of being a **singleton** relative to its storage, even if it's initialized in memory - so two levelup databases will share the same data-storage if initialized at the same storage location!
Now, ethereumjs uses levelup a lot for storage and caching and always initializes at storage location `''`, so there are sharing the same data. If now the cache is cleared (or a snapshot reverted) the cached data is removed, **it removes the data from all storages, since they are pointing to the same location**. If now another storage requesting the same data from a trie - it won't find it although the trie may points to it which finally leads to this error.

In my Implementation I build in a workaround for this by initializing each storage at a random location. This removed this bug in the most cases for me.
However, there is another _bug_ which is harder to resolve - memdown uses red-black-trees as a data-structure with a custom comparison function. The Trie stores all keys in the db as Buffers. If now the db is used for storing data outside of the trie, with other key types - e.g. Strings or Numbers, the custom comparison function has some weird behavior which i didn't debugged fully - it stores the keys in another location in the avl tree from where it tries to find and access it (I think different comparison functions are used), leading to a similar error to the showed one. This can be resolved by storing all keys as buffers.

With both solutions I didn't run into those bug anymore, however, there are different errors caused by a race condition by performing opening/closing and simultaneous read/write operation, which I couldn't solve, yet.

**Comment by tcoulter** [comment]
This should be fixed on the `persistence` branch: https://github.com/ethereumjs/testrpc/pull/275

TL;DR: When implementing persistence, I somehow was able to make this bug spring up quite regularly. When I moved everything to disk rather than in memory, the bug completely went away. I can't guarantee it's not still there, just as I can't disprove there's a flying 🦄 on the shady side of the 🌔, but the stark contrast in stability looks very promising.

I'm going to close this ticket for now, although it won't be _completely_ fixed until a new released is pushed. Keep a look out for the next version (4.0.0) which should be out soon.
